crates/sui-core/src/authority/authority_store.rs

```diff
@@ -610,43 +667,62 @@ impl AuthorityStore {
                         } else {
                             LockMode::ReadOnly
                         };
-                        (InputKey(*id, Some(*version)), lock_mode)
+                        (InputKey::Versioned{ id: *id, version: *version}, lock_mode)
                     }
                     // TODO: use ReadOnly lock?
-                    InputObjectKind::MovePackage(id) => (InputKey(*id, None), LockMode::Default),
+                    InputObjectKind::MovePackage(id) => (InputKey::Package { id: *id }, LockMode::Default),
                     // Cannot use ReadOnly lock because we do not know if the object is immutable.
-                    InputObjectKind::ImmOrOwnedMoveObject(objref) => (InputKey(objref.0, Some(objref.1)), LockMode::Default),
+                    InputObjectKind::ImmOrOwnedMoveObject(objref) => (InputKey::Versioned {id: objref.0, version: objref.1}, LockMode::Default),
                 }
             })
             .collect()
     }
 
     /// Checks if the input object identified by the InputKey exists, with support for non-system
-    /// packages i.e. when version is None.
-    pub fn multi_input_objects_exist(
+    /// packages i.e. when version is None. If the input object doesn't exist and it's a receiving
+    /// object, we also check if the object exists in the object marker table and view it as
+    /// existing if it is in the table.
+    pub fn multi_input_objects_available(
         &self,
         keys: impl Iterator<Item = InputKey> + Clone,
+        epoch_id: EpochId,
     ) -> Result<Vec<bool>, SuiError> {
-        let (keys_with_version, keys_without_version): (Vec<_>, Vec<_>) =
-            keys.enumerate().partition(|(_, key)| key.1.is_some());
+        let (keys_with_version, keys_without_version): (Vec<_>, Vec<_>) = keys
+            .enumerate()
+            .partition(|(_, key)| key.version().is_some());
 
-        let versioned_results = keys_with_version.iter().map(|(idx, _)| *idx).zip(
+        let mut versioned_results = vec![];
+        for ((idx, input_key), obj_wrapper_opt) in keys_with_version.iter().zip(
             self.perpetual_tables
                 .objects
                 .multi_get(
                     keys_with_version
                         .iter()
-                        .map(|(_, k)| ObjectKey(k.0, k.1.unwrap())),
+                        .map(|(_, k)| ObjectKey(k.id(), k.version().unwrap())),
                 )?
-                .into_iter()
-                .map(|o| o.is_some()),
-        );
+                .into_iter(),
+        ) {
+            match obj_wrapper_opt {
```

Comment:
it looks like we can use `multi_contains_keys` above, since we ignore the object value.